File: solutions/python/robot-name/1/robot_name.py
import random
import string

class Robot:
    _used_names = set()

    def __init__(self):
        self._name = self._generate_name()

    @property
    def name(self):
        return self._name

    def reset(self):
        self._name = self._generate_name()

    @classmethod
    def _generate_name(cls):
        while True:
            name = (
                random.choice(string.ascii_uppercase) +
                random.choice(string.ascii_uppercase) +
                str(random.randint(0, 9)) +
                str(random.randint(0, 9)) +
                str(random.randint(0, 9))
            )
            if name not in cls._used_names:
                cls._used_names.add(name)
                return name

File: solutions/javascript/robot-name/1/robot-name.js
const usedNames = new Set();

function generateName() {
  const letters = Array.from({ length: 2 }, () =>
    String.fromCharCode(65 + Math.floor(Math.random() * 26))
  ).join('');
  const digits = Array.from({ length: 3 }, () =>
    Math.floor(Math.random() * 10)
  ).join('');
  return `${letters}${digits}`;
}

function generateUniqueName() {
  let name;
  do {
    name = generateName();
  } while (usedNames.has(name));
  usedNames.add(name);
  return name;
}

export class Robot {
  get name() {
    if (!this._name) {
      this._name = generateUniqueName();
    }
    return this._name;
  }

  reset() {
    this._name = undefined;
  }
}

Robot.releaseNames = () => {
  usedNames.clear();
};